File: src/core/governance.ts
import type { Ledger } from "./ledger.js";
import type {
  DispatchRequest,
  DispatchResult,
  GovernanceConfig,
  GovernanceHandle,
  ProactivityStore,
} from "./types.js";

// Recursively sort object keys so equal targets serialize identically
// regardless of key order. A replacer-array on JSON.stringify won't do this —
// it's an allowlist applied at every depth, so nested keys not present at the
// top level get dropped and distinct nested targets collide.
const canonicalize = (value: unknown): unknown =>
  value && typeof value === "object" && !Array.isArray(value)
    ? Object.fromEntries(
        Object.keys(value as Record<string, unknown>)
          .sort()
          .map((k) => [k, canonicalize((value as Record<string, unknown>)[k])]),
      )
    : value;

// Canonical key for an action: same action + target + tick collapses to one
// attempt, so a re-run within a tick is idempotent rather than double-fired.
const deriveIdempotencyKey = (parts: {
  actionType: string;
  target: Record<string, unknown>;
  tickId: string;
}): string =>
  `${parts.actionType}:${JSON.stringify(canonicalize(parts.target))}:${parts.tickId}`;

export const createGovernance = (
  config: GovernanceConfig,
  tickId: string,
  entityId: string,
  ledger: Ledger,
): GovernanceHandle => {
  const { store, caps, softCaps = [], dryRun = false } = config;

  const dispatch = async (request: DispatchRequest): Promise<DispatchResult> => {
    const idempotencyKey = deriveIdempotencyKey({
      actionType: request.actionType,
      target: request.target,
      tickId,
    });

    const tickActionCount = ledger.countActionsTaken();
    if (tickActionCount >= caps.perTick) {
      return await recordDenied(store, {
        ...request,
        tickId,
        idempotencyKey,
        outcome: "hard_denied",
        denialReason: `Per-tick cap reached (${caps.perTick})`,
      });
    }

    const passActionCount = ledger.countActionsForPass(request.goalTickId);
    if (passActionCount >= caps.perPass) {
      return await recordDenied(store, {
        ...request,
        tickId,
        idempotencyKey,
        outcome: "hard_denied",
        denialReason: `Per-pass cap reached (${caps.perPass})`,
      });
    }

    for (const sc of softCaps) {
      const result = sc.evaluate({
        actionType: request.actionType,
        target: request.target,
        recentAttempts: await store.getRecentAttempts(entityId, { tickWindow: 5 }),
      });
      if (result.triggered && !request.overrideReason) {
        return await recordDenied(store, {
          ...request,
          tickId,
          idempotencyKey,
          outcome: "soft_cap_denied",
          denialReason:
            result.warning ??
            `Soft cap "${sc.name}" triggered — supply an overrideReason to proceed`,
        });
      }
    }

    const insertResult = await store.insertAttempt({
      goalId: request.goalId,
      tickId,
      goalTickId: request.goalTickId,
      actionType: request.actionType,
      idempotencyKey,
      governanceOutcome: dryRun ? "pending_approval" : "taken",
      reasoning: request.reasoning,
      denialReason: null,
      overrideReason: request.overrideReason ?? null,
      target: request.target,
      payload: request.payload ?? null,
    });

    if (insertResult.kind === "idempotency_conflict") {
      return {
        governanceOutcome: "hard_denied",
        attemptId: insertResult.prior.attemptId,
        idempotencyKey,
        denialReason: `Duplicate: prior attempt ${insertResult.prior.attemptId}`,
      };
    }

    const { attemptId } = insertResult;

    if (dryRun) {
      ledger.record({
        goalId: request.goalId,
        goalTickId: request.goalTickId,
        actionType: request.actionType,
        outcome: "pending_approval",
      });
      return { governanceOutcome: "pending_approval", attemptId, idempotencyKey };
    }

    try {
      await request.perform();
      await store.markAttemptCompleted(attemptId);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      await store.markAttemptFailed(attemptId, message);
      return {
        governanceOutcome: "hard_denied",
        attemptId,
        idempotencyKey,
        denialReason: `Side effect failed: ${message}`,
      };
    }

    const outcome = request.overrideReason ? "soft_cap_overridden" : "taken";
    ledger.record({
      goalId: request.goalId,
      goalTickId: request.goalTickId,
      actionType: request.actionType,
      outcome,
    });

    return {
      governanceOutcome: outcome,
      attemptId,
      idempotencyKey,
      overrideReason: request.overrideReason,
    };
  };

  return { dispatch };
};

const recordDenied = async (
  store: ProactivityStore,
  opts: DispatchRequest & {
    tickId: string;
    idempotencyKey: string;
    outcome: "hard_denied" | "soft_cap_denied";
    denialReason: string;
  },
): Promise<DispatchResult> => {
  // The audit row gets a suffixed key: only delivered/pending attempts may
  // claim the canonical key. If a denial claimed it, a soft_cap_denied action
  // re-dispatched with an overrideReason would collide with its own denial row
  // and be rejected as a duplicate — the documented retry path would never work.
  const result = await store.insertAttempt({
    goalId: opts.goalId,
    tickId: opts.tickId,
    goalTickId: opts.goalTickId,
    actionType: opts.actionType,
    idempotencyKey: `${opts.idempotencyKey}#denied:${crypto.randomUUID()}`,
    governanceOutcome: opts.outcome,
    reasoning: opts.reasoning,
    denialReason: opts.denialReason,
    overrideReason: null,
    target: opts.target,
    payload: opts.payload ?? null,
  });

  const attemptId = result.kind === "inserted" ? result.attemptId : result.prior.attemptId;

  return {
    governanceOutcome: opts.outcome,
    attemptId,
    idempotencyKey: opts.idempotencyKey,
    denialReason: opts.denialReason,
  };
};

File: integrations/hermes/proactivity_hermes/governance.py
"""The governance envelope — a Python port of `@refix/proactivity`'s
`core/governance.ts`, adapted for Hermes.

This is the crown jewel: every outbound action the agent attempts passes
through `dispatch`, which enforces idempotency, a per-tick action cap, and
pluggable soft caps before the side effect runs, and records an audit trail.

It depends only on a store that implements the small attempt-protocol below
(`count_taken_for_tick`, `get_recent_attempts`, `insert_attempt`,
`mark_attempt_completed`, `mark_attempt_failed`) — `SqliteStore` provides it,
but the logic is store-agnostic and unit-testable on its own.

Differences from the TS original, by design:
- No in-memory `Ledger`. In Hermes each governed tool call is an independent
  middleware invocation, so there is no shared per-tick ledger to thread —
  the per-tick count is read from the store.
- No per-pass cap / `goal_tick`. We skip the plan/act split for a personal
  agent (Hermes drives its own loop), so the only cap is per-tick.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Callable, Optional, Sequence


def idempotency_key(action_type: str, target: dict, tick_id: str) -> str:
    """A canonical key for an action: same action + target + tick collapse to
    one attempt, so a re-run within a tick is idempotent rather than double-fired.

    `sort_keys=True` canonicalizes nested dict key order (array order is kept,
    since it is semantic) — the stdlib equivalent of the TS `canonicalize`.
    """
    body = json.dumps(target, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    return f"{action_type}:{body}:{tick_id}"


# A soft cap inspects the action plus recent attempts and returns a warning
# string to hold the action, or None to let it through. `override_reason`
# bypasses all soft caps.
SoftCap = Callable[[str, dict, Sequence[Any]], Optional[str]]


@dataclass
class DispatchRequest:
    action_type: str
    target: dict
    perform: Callable[[], Any]  # runs the real side effect; its return is surfaced
    payload: Any = None
    reasoning: str = ""
    override_reason: Optional[str] = None
    goal_id: Optional[str] = None


@dataclass
class DispatchResult:
    outcome: str  # taken | soft_cap_overridden | hard_denied | pending_approval
    performed: bool
    result: Any = None  # the side effect's own output, when performed
    denial_reason: Optional[str] = None
    attempt_id: Optional[str] = None


class Governance:
    def __init__(
        self,
        store,
        *,
        entity_id: str,
        per_tick_cap: int,
        soft_caps: Optional[Sequence[SoftCap]] = None,
        recent_window: int = 5,
        dry_run: bool = False,
    ):
        self._store = store
        self._entity_id = entity_id
        self._per_tick_cap = per_tick_cap
        self._soft_caps = list(soft_caps or [])
        self._recent_window = recent_window
        self._dry_run = dry_run

    def dispatch(self, tick_id: str, req: DispatchRequest) -> DispatchResult:
        key = idempotency_key(req.action_type, req.target, tick_id)

        # 1. Per-tick cap. Counted from the store because each governed call is
        # an independent middleware invocation — there is no shared ledger.
        if self._store.count_taken_for_tick(self._entity_id, tick_id) >= self._per_tick_cap:
            return self._deny(tick_id, key, req, f"Per-tick cap reached ({self._per_tick_cap})")

        # 2. Soft caps (skipped when the caller supplied an override reason).
        if not req.override_reason:
            recent = self._store.get_recent_attempts(self._entity_id, self._recent_window)
            for cap in self._soft_caps:
                warning = cap(req.action_type, req.target, recent)
                if warning:
                    return self._deny(tick_id, key, req, warning)

        # 3. Reserve the attempt. The UNIQUE idempotency_key turns a duplicate
        # into a conflict, which is how re-runs are deduped.
        outcome = (
            "pending_approval"
            if self._dry_run
            else ("soft_cap_overridden" if req.override_reason else "taken")
        )
        ins = self._store.insert_attempt(
            entity_id=self._entity_id,
            tick_id=tick_id,
            action_type=req.action_type,
            idempotency_key=key,
            governance_outcome=outcome,
            target=req.target,
            payload=req.payload,
            goal_id=req.goal_id,
            reasoning=req.reasoning,
            override_reason=req.override_reason,
        )
        if not ins.inserted:
            return DispatchResult(
                outcome="hard_denied",
                performed=False,
                attempt_id=ins.attempt_id,
                denial_reason=f"Duplicate of prior attempt {ins.attempt_id}",
            )

        # 4. Dry run stops before the side effect, leaving a pending row to approve.
        if self._dry_run:
            return DispatchResult(outcome="pending_approval", performed=False, attempt_id=ins.attempt_id)

        # 5. Perform the real action. Any error is caught and recorded — it never
        # escapes dispatch, so a caller's fail-open path cannot double-fire.
        try:
            result = req.perform()
        except Exception as exc:  # noqa: BLE001 — record and surface, never leak
            self._store.mark_attempt_failed(ins.attempt_id, str(exc))
            return DispatchResult(
                outcome="hard_denied",
                performed=False,
                attempt_id=ins.attempt_id,
                denial_reason=f"Side effect failed: {exc}",
            )
        self._store.mark_attempt_completed(ins.attempt_id)
        return DispatchResult(outcome=outcome, performed=True, result=result, attempt_id=ins.attempt_id)

    def _deny(self, tick_id: str, key: str, req: DispatchRequest, reason: str) -> DispatchResult:
        ins = self._store.insert_attempt(
            entity_id=self._entity_id,
            tick_id=tick_id,
            action_type=req.action_type,
            idempotency_key=key,
            governance_outcome="hard_denied",
            target=req.target,
            payload=req.payload,
            goal_id=req.goal_id,
            reasoning=req.reasoning,
            denial_reason=reason,
        )
        return DispatchResult(
            outcome="hard_denied", performed=False, attempt_id=ins.attempt_id, denial_reason=reason
        )
